### DISCUSSION TITLE
drone stability

### DISCUSSION THREAD

**Discussion by dtfx** [bitcraze/discussions#370] [Q&A] [answered]
hello, I'm building a drone and I want my drone to stay stable in the air. I am using PWM3901 and VL53L1X sensor for this. How can I do this project only this part. Can we convert the value of PWM3901 to millimeters or centimeters?

https://github.com/bitcraze/Bitcraze_PMW3901?utm_source=rb-community&utm_medium=forum&utm_campaign=pmw3901-sensor-for-legged-robot-position-estimation

Is the above code only reading the data from the sensor?
How can I combine this data with elevation data ?

**Comment by ataffanel** [comment, marked as answer]
The data you get from the PWM3901 are raw "pixel" data: this sensor is a fast low resolution camera with optical flow and you literally get how accumulated pixels count the image has likely moved on the sensor.

There is, as far as I know, no ready-to-use code to interpret these data as physical movement. However you can have a look at the [Crazyflie kalman implementation](https://github.com/bitcraze/crazyflie-firmware/blob/68a7ec98ff1b93a2a0d3460819bb7e97e797657f/src/modules/src/kalman_core/mm_flow.c#L35). Note that this code is doing the oposite of what you want: it calculates the number of pixel from a predicted velocity and compare it to the current measurement in pixels (this is how data are pushed in a Kalman filter).

Depending of your goals you could either try to invert the formulas of try to use a Kalman filter in the same way.

**Reply by dtfx** [reply]
Many thanks for the information

**Reply by knmcguire** [reply]
If you are satisfied with the answer, could you select 'mark as answer' please? Thanks!
